config/src/config.rs

```diff
@@ -0,0 +1,292 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use bigdecimal::{BigDecimal, FromPrimitive, ToPrimitive};
+use figment::{
+    providers::{Env, Format, Toml},
+    Figment,
+};
+use serde_repr::Deserialize_repr;
+use serde_with::DurationSecondsWithFrac;
+use std::{collections::HashMap, net::SocketAddr, path::PathBuf, str::FromStr, time::Duration};
+use tracing::warn;
+
+use alloy_primitives::Address;
+use bip39::Mnemonic;
+use serde::Deserialize;
+use serde_with::serde_as;
+use thegraph::types::DeploymentId;
+use url::Url;
+
+use crate::NonZeroGRT;
+
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct Config {
+    pub indexer: IndexerConfig,
+    pub database: DatabaseConfig,
+    pub graph_node: GraphNodeConfig,
+    pub metrics: MetricsConfig,
+    pub subgraphs: SubgraphsConfig,
+    pub blockchain: BlockchainConfig,
+    pub service: ServiceConfig,
+    pub tap: TapConfig,
+}
+
+pub enum ConfigPrefix {
+    Tap,
+    Service,
+}
+
+impl ConfigPrefix {
+    fn get_prefix(&self) -> &'static str {
+        match self {
+            Self::Tap => "TAP_AGENT_",
+            Self::Service => "INDEXER_SERVICE_",
+        }
+    }
+}
+
+impl Config {
+    pub fn parse(prefix: ConfigPrefix, filename: &PathBuf) -> Result<Self, String> {
+        let config_defaults = include_str!("../default_values.toml");
+
+        let config: Self = Figment::new()
+            .merge(Toml::string(config_defaults))
+            .merge(Toml::file(filename))
+            .merge(Env::prefixed(prefix.get_prefix()))
+            .extract()
+            .map_err(|e| e.to_string())?;
+        config.validate()?;
+
+        Ok(config)
+    }
+
+    // custom validation of the values
+    fn validate(&self) -> Result<(), String> {
+        let one = BigDecimal::from(1);
+        let ten = BigDecimal::from(10);
+        match &self.tap.rav_request.trigger_value_divisor {
+            x if x <= &one => {
+                return Err("trigger_value_divisor must be greater than 1".to_string())
+            }
+            x if x > &one && x < &ten => warn!(
```

Comment:
Wouldn't it be simpler to do this?
```suggestion
            x if *x > 1.into() && *x < 10.into() => warn!(
```

Reply:
Thanks for that. I updated that part of the code, but multiplication wasn't possible to do it because of into().